File: backend-bot/src/azure-storage/azure-storage.service.ts
import { Injectable } from '@nestjs/common';
import {
  BlobServiceClient,
  ContainerClient,
  BlockBlobClient,
  BlockBlobUploadResponse,
  BlobDeleteIfExistsResponse,
  StorageSharedKeyCredential,
  generateAccountSASQueryParameters,
  AccountSASSignatureValues,
  AccountSASPermissions,
  AccountSASResourceTypes,
  AccountSASServices,
  BlockBlobUploadOptions,
} from '@azure/storage-blob';
@Injectable()
export class AzureStorageService {
  private blobServiceClient: BlobServiceClient;
  private conectionString: string;
  private accountName: string;
  private accountKey: string;
  private sharedKeyCredential: StorageSharedKeyCredential;

  constructor() {
    this.conectionString = process.env.AZURE_STORAGE_CONNECTION_STRING;
    this.accountName = process.env.AZURE_STORAGE_ACCOUNT_NAME;
    this.accountKey = process.env.AZURE_STORAGE_ACCOUNT_KEY;
    this.blobServiceClient = BlobServiceClient.fromConnectionString(this.conectionString);
    this.sharedKeyCredential = new StorageSharedKeyCredential(this.accountName, this.accountKey);
  }

  async upload(
    file: Express.Multer.File,
    containerName: string,
    fileName: string,
  ): Promise<BlockBlobUploadResponse> {
    try {
      const blobClient: BlockBlobClient = await this.getBlobClient(containerName, fileName);
      // Define las opciones de carga del Blob, incluyendo el Content-Type
      const uploadOptions: BlockBlobUploadOptions = {
        blobHTTPHeaders: {
          blobContentType: file.mimetype,
        },
      };

      const uploadFile = file.buffer;
      return blobClient.upload(uploadFile, file.size, uploadOptions);
    } catch (err) {
      console.log(err);
      throw err;
    }
  }

  async delete(containerName: string, fileName: string): Promise<BlobDeleteIfExistsResponse> {
    try {
      const blobClient: BlockBlobClient = await this.getBlobClient(containerName, fileName);
      return blobClient.deleteIfExists();
    } catch (err) {
      console.log(err);
    }
  }

  async getUrl(containerName: string, fileName: string): Promise<string> {
    try {
      const blobClient: BlockBlobClient = await this.getBlobClient(containerName, fileName);
      return blobClient.url;
    } catch (err) {
      console.log(err);
    }
  }

  async getSasUrl(containerName: string, fileName: string): Promise<string> {
    try {
      const blobClient: BlockBlobClient = await this.getBlobClient(containerName, fileName);
      return `${blobClient.url}?${this.getBlobSas()}`;
    } catch (err) {
      console.log(err);
    }
  }

  private getBlobSas(): string {
    const tokenValidDuration = 9 * 60 * 1000; // 9 minutes
    const adjustTime = 60 * 1000; // 1 minute for potential time discrepancies
    const blobSAS: AccountSASPermissions = AccountSASPermissions.parse('r');
    const sasOptions: AccountSASSignatureValues = {
      permissions: blobSAS,
      startsOn: new Date(new Date().valueOf() - adjustTime),
      expiresOn: new Date(new Date().valueOf() + tokenValidDuration),
      services: AccountSASServices.parse('bf').toString(), // blobs, tables, queues, files
      resourceTypes: AccountSASResourceTypes.parse('sco').toString(), // service, container, object
    };
    return generateAccountSASQueryParameters(sasOptions, this.sharedKeyCredential).toString();
  }

  private async getBlobClient(containerName: string, fileName: string): Promise<BlockBlobClient> {
    const containerClient: ContainerClient = this.blobServiceClient.getContainerClient(containerName);

    try {
      await containerClient.createIfNotExists();
    } catch (err) {
      console.log(err);
    }
    return containerClient.getBlockBlobClient(fileName);
  }

  async existBlob(containerName: string, fileName: string): Promise<boolean> {
    const blobClient: BlockBlobClient = await this.getBlobClient(containerName, fileName);
    return blobClient.exists();
  }

  async getBinary(containerName: string, fileName: string): Promise<Buffer> {
    const blobClient: BlockBlobClient = await this.getBlobClient(containerName, fileName);
    const downloadBlockBlobResponse = await blobClient.download();
    return await this.streamToBuffer(downloadBlockBlobResponse.readableStreamBody);
  }

  private async streamToBuffer(readableStream: NodeJS.ReadableStream): Promise<Buffer> {
    return new Promise<Buffer>((resolve, reject) => {
      const chunks: Uint8Array[] = [];
      readableStream.on('data', (data: Uint8Array) => {
        chunks.push(data);
      });
      readableStream.on('end', () => {
        resolve(Buffer.concat(chunks));
      });
      readableStream.on('error', reject);
    });
  }
}

File: backend-doc-analyzer/src/services/azure_storage_service.py
from datetime import datetime, timedelta
from os import getenv
from typing import IO, Any, AnyStr, Dict, Iterable, List

from src.entities.file_entity import FileEntity
from azure.storage.blob import (
    BlobServiceClient,
    BlobClient,
    generate_blob_sas,
    AccountSasPermissions,
    ResourceTypes,
    ContentSettings,
)
from src.config import AZURE_STORAGE_CONNECTION_STRING
from src.decorators.map_exceptions import map_exceptions, exceptions_mapped
from src.exceptions.storage.service_unavailable_exception import (
    ServiceUnavailableException,
)


@map_exceptions(ServiceUnavailableException("Blob Storage Error"))
class AzureStorageService(object):
    @exceptions_mapped
    def __init__(self):
        self._connection_string = AZURE_STORAGE_CONNECTION_STRING
        self._blob_service_client = BlobServiceClient.from_connection_string(
            self._connection_string
        )

    @exceptions_mapped
    def upload(
        self,
        file_name: str,
        file: bytes | str | Iterable[AnyStr] | IO[AnyStr],
        container_name: str,
        content_type: str = None,
    ) -> Dict[str, Any]:
        blob_client = self._get_blob_client(file_name, container_name)
        self.delete(file_name, container_name)
        content_settings = ContentSettings(content_type=content_type)
        blob_client.upload_blob(data=file, content_settings=content_settings)

    @exceptions_mapped
    def delete(self, file_name: str, container_name: str) -> bool:
        blob_client = self._get_blob_client(file_name, container_name)
        if blob_client.exists():
            blob_client.delete_blob(delete_snapshots="include")
            return True
        return False

    @exceptions_mapped
    def get(self, file_name: str, container_name: str) -> bytes:
        blob_client = self._get_blob_client(file_name, container_name)
        return blob_client.download_blob().readall()
    @exceptions_mapped
    def get_content_type(self, file_name: str, container_name: str) -> str:
        blob_client = self._get_blob_client(file_name, container_name)
        return blob_client.get_blob_properties().content_settings.content_type

    @exceptions_mapped
    def get_url(self, file_name: str, container_name: str) -> str:
        return f"{self._blob_service_client.url}{container_name}/{file_name}"

    @exceptions_mapped
    def get_sas_url(self, file_name: str, container_name: str) -> str:
        return f"{self.get_url(file_name, container_name)}?{self._get_blob_sas(file_name, container_name)}"

    def _get_blob_sas(self, file_name: str, container_name: str) -> str:
        blob_client = self._get_blob_client(file_name, container_name)
        return generate_blob_sas(
            blob_client.account_name,
            container_name,
            file_name,
            account_key=blob_client.credential.account_key,
            resource_types=ResourceTypes(object=True),
            permission=AccountSasPermissions(read=True),
            start=datetime.utcnow() - timedelta(minutes=15),
            expiry=datetime.utcnow() + timedelta(minutes=60),
        )
    
    def get_sas_url_group(self, files: List[FileEntity], container_name: str) -> List[str]:
        urls = []
        for file in files:
            url = self.get_sas_url(file_name=str(object=file.id), container_name=container_name)
            document_info = {
                "id": str(file.id),
                "filename": file.display_name,
                "url": url
            }
            urls.append(document_info)
        return urls

    def _get_blob_client(self, file_name: str, container_name: str) -> BlobClient:
        container_client = self._blob_service_client.get_container_client(
            container_name
        )
        if not container_client.exists():
            container_client.create_container()
        blob_client = container_client.get_blob_client(file_name)
        return blob_client
    